src/Featurizers/Components/DocumentStatisticsEstimator.h

```diff
@@ -240,23 +298,182 @@ inline DocumentStatisticsAnnotationData::DocumentStatisticsAnnotationData(Freque
-// |  Details::DocumentStatisticsTrainingOnlyPolicy
+// |  DocumentStatisticsEstimator
 // |
 // ----------------------------------------------------------------------
-inline Details::DocumentStatisticsTrainingOnlyPolicy::DocumentStatisticsTrainingOnlyPolicy(void) :
-    _totalNumDocuments(0) {
+//Would it be better to place default values during pipeline chaining(countvectorizer and tfidfvectorizer)? if so this ctor will show no default values
+template <size_t MaxNumTrainingItemsV>
+DocumentStatisticsEstimator<MaxNumTrainingItemsV>::DocumentStatisticsEstimator(
+    AnnotationMapsPtr pAllColumnAnnotations, 
+    size_t colIndex,
+    StringDecorator decorator, 
+    nonstd::optional<IndexMap> existingVocabulary, 
+    nonstd::optional<std::uint32_t> topKTerms, 
+    std::float_t minDf, 
+    std::float_t maxDf
+) :
+    BaseType(
+        std::move(pAllColumnAnnotations),
+        std::move(colIndex),
+        true,
+        std::move(decorator),
+        std::move(existingVocabulary),
+        std::move(topKTerms),
+        std::move(minDf),
+        std::move(maxDf)
+    ) {
 }
 
-inline Details::DocumentStatisticsTrainingOnlyPolicy::DocumentStatisticsTrainingOnlyPolicy(StringDecorator decorator) :
+// ----------------------------------------------------------------------
+// |
+// |  Anonymous namespace
+// |
+// ----------------------------------------------------------------------
+namespace {
+
+inline Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap PruneTermFreqMap(Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap & termFrequency, 
+                                                                                    std::float_t minDf, 
+                                                                                    std::float_t maxDf, 
+                                                                                    std::float_t totalNumDocumentsFloat,
+                                                                                    nonstd::optional<std::uint32_t> topKTerms) {
+
+    using FrequencyMap = Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap;
+    
+    //trim by minDf and maxDf
+    if (minDf > 0.0f || maxDf < 1.0f) {
+        for (FrequencyMap::iterator termFrequencyIter = termFrequency.begin(); termFrequencyIter != termFrequency.end();) {
+            std::float_t const freq = termFrequencyIter->second / totalNumDocumentsFloat;
+
+            if (freq < minDf || freq > maxDf) {
+                termFrequencyIter = termFrequency.erase(termFrequencyIter);
+            } else {
+                ++termFrequencyIter;
+            }       
+        }
+    }
+
+    //trim by topKTerms
+    if (!topKTerms.has_value() || termFrequency.empty())
+        return termFrequency;
+
+    std::uint32_t const & topKTermsValue(*topKTerms);
+
+    if (topKTermsValue > termFrequency.size())
+        throw std::invalid_argument("topKTermsUpperBound");
+
+    std::vector<std::uint32_t> frequencyVector;
```

Comment:
I'm concerned about this logic - I see how it works during a full retrain scenario, but not clear how it maintains values after multiple retrains. Here is the scenario that I am thinking about:

  1) Train, 100 terms are calculated (meaning ids 0-99 (or 1-100) are assigned)
  2) Incremental training with original 100 items, where 10 new words are discovered (meaning ids 0-109 (or 1-110) are assigned)
  3) User removes random words in the training set (for example, maybe those words aren't meaningful) <<<< This may be an invalid operation; if so, disregard the rest of this
  4) We retrain, and IndexMap will fill in the gaps (I just verified that there is a bug in IndexMap where it potentially assigns duplicate values in this scenario)

So, the problem is that we will remove some of the values in the incoming map when I believe that we should always treat those as more important than the new terms that we discovered during retraining.